### ISSUE TITLE
OptionalParametersAfterRequiredRector breaks code

### ISSUE THREAD

**Issue by dakujem** [rectorphp/rector#8950]
| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | `2.0.4`          |

I assumed the [`OptionalParametersAfterRequiredRector`](https://getrector.com/rule-detail/optional-parameters-after-required-rector) rule was wrong, but it is actually [working well](https://getrector.com/demo/8620157b-63b0-4172-a6b9-e15ec70f2b47) if it can infer the file types correctly.

The issue is that there are places where Recotr is not able to infer the type of the service and does not update the code properly - see here: https://getrector.com/demo/b0155317-b123-4480-8a21-70638dc97320

## Minimal PHP Code Causing Issue

https://getrector.com/demo/b0155317-b123-4480-8a21-70638dc97320

## Expected Behaviour

Alternative rule that would instead remove the `= null` default, or any default in that matter.

before:
```php
function foobar($optional = null, $required){}
```
after:
```php
function foobar($optional, $required){}
```

This would remove the PHP 8.4 deprecation warning, but keep the code 100% backwards compatible, because it is impossible to call the code without passing the first argument anyway ... 🤔 ... On th4e second thought, named arguments might break that assumption.

**Comment by samsonasik** [comment]
That's expected, the `foo()` doesn't have return type, you can add return type via typeDeclarations set, and it will correctly flipped.

because when you check on phpstan, it will got notice : call foobar on mixed

https://phpstan.org/r/b1f0b093-a2bf-43c8-8ef2-93cf792ef5f2

It of course, needs step by step process in case of usage by type returned from function

**First**: make typed

- https://getrector.com/demo/03e0f4d6-3ecb-4e48-8815-d92b9e1af4df

**Second**: enable `OptionalParametersAfterRequiredRector`

- https://getrector.com/demo/80cd67ff-fe8a-44b8-8e64-44ff1333c542

Apply PHPStan and Rector on tandem so you get full feature :)

take it slow to integrate to new project :), you can read step by step here https://getrector.com/documentation/integration-to-new-project

**Comment by dakujem** [comment]
I know I can add types in such a simple example. What about places like PSR DI container's `get` method?

`Env::container()->get(Service::class)->someMethod($optional, $required)`

Before I am hinted to modify `Env::container()` method's doc-comment, there are places where that is not an option, like when injecting into controllers via the PSR _interface_.
`public function endpoint(        \Psr\Container\ContainerInterface $container    ) {` ...

**Comment by dakujem** [comment]
anyway, why close the issue? is it not a valid idea to create a rule that does what I suggested above? 🤔

**Comment by samsonasik** [comment]
If you look on edge case, it will always an edge case :), and end up will not doing anything.

Remove default value is not expected, as it will remove the expected behaviour.

To integrate on legacy project, take it slow is the way to go.

**Comment by dakujem** [comment]
Well that's what I did - I updated the code by hand, but it would help to have such a rule/ruleset in rector, since the case I mentioned is widespread and I'm not sure how to tell rector that some cases simply return instances by requested class name in the parameters. One can do this with phpstorm easily. Would something like [this PHPstan extension `phpstan-psr-container`](https://github.com/bnf/phpstan-psr-container) help?
